### PR TITLE
Fixed the subscription tracker mixin

### PR DESCRIPTION
**Pull request by ryan-bradford**
# Description

Updates the subscription tracker mixin to be abstract and require a ngOnDestroy method. Also, changes the subscription tracker mixin example to have an empty implementation of ngOnDestroy.

# Testing

Ran the examples app in prod mode and made sure the mixin worked.

# Future Work

Creating tslint rules:
https://jira.eng.vmware.com/browse/VDUCC-140

Signed-off-by: Ryan Bradford rbradford@vmware.com

### CODE REVIEW

File: projects/components/src/common/subscription/subscription-tracker.mixin.ts [hunk 1]
```diff
@@ -35,8 +35,7 @@ export function SubscriptionTrackerMixin<TBase extends Type<{}>>(Base: TBase) {
             this.tracker.unsubscribeAll();
         }
 
-        ngOnDestroy(): void {
-            this.unsubscribeAll();
-        }
-    };
+        abstract ngOnDestroy(): void;
```
Comment (ps37) [inline projects/components/src/common/subscription/subscription-tracker.mixin.ts:38]: Why was this made abstract by removing the call to this.unsubscribeAll? Wouldn't that make all the subclasses duplicate the call to this.unsubscribeAll?

File: projects/components/src/datagrid/filters/datagrid-filter.ts [hunk 2]
```diff
@@ -93,6 +93,8 @@ export abstract class DatagridFilter<V, C extends FilterConfig<V>> extends Subsc
         this.subscribe(obs, () => this.changes.next());
     }
 
+    ngOnDestroy(): void {}
```
Comment (ps37) [inline projects/components/src/datagrid/filters/datagrid-filter.ts:96]: If we go down the route of making this abstract, ain't we supposed to call unsubscribeAll here?

File: projects/examples/src/components/subscription/subscription-tracker-mixin.example.component.ts [hunk 3]
```diff
@@ -18,13 +18,16 @@ import { BehaviorSubject, Observable } from 'rxjs';
...
+export class SubscriptionTrackerMixinExampleSubComponent extends SubscriptionTrackerMixin(class {})
+    implements OnInit, OnDestroy {
     @Input()
     observable: Observable<number>;
 
     ngOnInit(): void {
         this.subscribe(this.observable, value => console.log(value));
     }
+
+    ngOnDestroy(): void {}
```
Comment (ps37) [inline projects/examples/src/components/subscription/subscription-tracker-mixin.example.component.ts:30]: Same here

**Comment by juanmendes** [comment]
It seems the benefits of mixins that require returning abstract classes are made impossible by true support from AOT and TS itself.

I think the best thing to do is to stop using the SubscriptionTracker mixin. It seems that mixins outputting an abstract class is not supported because typescript loses the abstractness of the class it returns when included from a .d.ts  It works fine within a project but doesn’t work when included from node_modules.

Angular also requires that you create an ngOnDestroy on your object otherwise it won’t call  ngOnDestroy in AOT mode. It works fine for regular base classes.
With all this, I’m proposing we should not make use of mixins for abstract base classes,  e.g., subscription tracker. As much as I love the pattern 

Please remove the mixin.

image

File: projects/components/src/datagrid/filters/datagrid-filter.ts (hunk 2 above)
Comment (ryan-bradford) [reply projects/components/src/datagrid/filters/datagrid-filter.ts:96]: The logic of calling unsubscribeAll is handled by SubscriptionTracker in the constructor. It modifies the ngOnDestroy method itself.

File: projects/components/src/common/subscription/subscription-tracker.mixin.ts (hunk 1 above)
Comment (ryan-bradford) [reply projects/components/src/common/subscription/subscription-tracker.mixin.ts:38]: Removed in the most recent version, but the logic to call unsubscribeAll is in the constructor of the SubscriptionTracker

**Comment by ryan-bradford** [comment]
Just added a commit to remove all mixins, rebased to origin/master, and squashed.

**Pull request merged**